homeassistant/components/powerwall/sensor.py

```diff
@@ -0,0 +1,159 @@
+        return False
+
+    async def async_update(self):
+        """Update the entity.
+
+        Only used by the generic entity update service.
+        """
+        await self._coordinator.async_request_refresh()
+
+    async def async_added_to_hass(self):
+        """Subscribe to updates."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
+
+    async def async_will_remove_from_hass(self):
+        """Undo subscription."""
+        self._coordinator.async_remove_listener(self.async_write_ha_state)
+
+
+class PowerWallChargeSensor(PowerWallSensor):
+    """Representation of an Powerwall charge sensor."""
+
+    def __init__(self, coordinator, site_info, ip_address):
+        """Initialize the sensor."""
+        super().__init__(coordinator, site_info, ip_address)
+        self._coordinator = coordinator
+        self._site_info = site_info
+        self._ip_address = ip_address
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return "%"  # switch to UNIT_PERCENTAGE
+
+    @property
+    def name(self):
+        """Device Name."""
+        return "Powerwall Charge"
+
+    @property
+    def device_class(self):
+        """Device Class."""
+        return DEVICE_CLASS_BATTERY
+
+    @property
+    def unique_id(self):
+        """Device Uniqueid."""
+        return f"{self._ip_address}_charge"
+
+    @property
+    def state(self):
+        """Get the current value in percentage."""
+        return round(self._coordinator.data[POWERWALL_API_CHARGE], 3)
+
+
+class PowerWallEnergySensor(PowerWallSensor):
+    """Representation of an Powerwall Energy sensor."""
+
+    def __init__(self, meter, coordinator, site_info, ip_address):
+        """Initialize the sensor."""
+        super().__init__(coordinator, site_info, ip_address)
+        self._coordinator = coordinator
+        self._meter = meter
+        self._site_info = site_info
+        self._ip_address = ip_address
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return "kWh"
+
+    @property
+    def name(self):
+        """Device Name."""
+        return f"Powerwall {self._meter.title()} Now"
+
+    @property
+    def device_class(self):
+        """Device Class."""
+        return DEVICE_CLASS_POWER
+
+    @property
+    def unique_id(self):
+        """Device Uniqueid."""
+        return f"{self._ip_address}_{self._meter}_instant_power"
+
+    @property
+    def state(self):
+        """Get the current value in kWh."""
+        meter = self._coordinator.data[POWERWALL_API_METERS][self._meter]
+        return round(float(meter.instant_power / 1000), 3)
+
+    @property
+    def device_state_attributes(self):
+        """Return the device specific state attributes."""
+        meter = self._coordinator.data[POWERWALL_API_METERS][self._meter]
+        return {
+            ATTR_FREQUENCY: meter.frequency,
```

Comment:
What does the frequency measurement report? What kind of value is it?

Reply:
Its the power frequency in Hz